File: CodiceFiscale/model/sesso.cpp
#include "sesso.h"
#include <limits>

Sesso::Sesso() : genere('0') {}


Sesso::Sesso(const Sesso& x) {
    genere = x.genere;
}


char Sesso::estraiGenere() const throw() {
    return genere;
}

Sesso& Sesso::operator=(const char& x) {
    if(genere != x) genere=x;
    return *this;
}

std::istream& operator>>(std::istream& is, Sesso& s) throw(CFexception) {
    char c;
    std::cout<<"Inserire sesso (digitare m oppure f): ";
    if(!(is >> c)) throw FineFileSesso();
    s.genere = c;
    return is;
}


void Sesso::creaDaCodice (const std::string& cod) throw() {
    //validita' data (e quindi sesso) controllata nei precedenti controlli su Data

    //estrazione codice giorno da codice fiscale (sulla base del quale e' calcolato il sesso)
    int n = stringToUInt(cod);
    if(n>0 && n<=31) genere = 'M';
        else genere = 'F'; //da metodo Nascita::creaDaCodice, l'altro intervallo è necessariamente n>40&&n<=71
}


std::string Sesso::generaCodice() const throw() {
    //valori possibili per genere: f/F/m/M
    if(genere == 'f' || genere == 'F') return "40";
    else return "0";
}


void Sesso::controllo() throw(CFexception) {
    if(genere!='m' && genere!='f' && genere!='M' && genere!='F') throw ErrSesso();
}


std::string Sesso::toString() const throw() {
    //valori possibili per genere: f/F/m/M
    if(genere == 'M' || genere == 'm') return "MASCHIO";
    else return "FEMMMINA";
}


Sesso::operator bool() const throw() {
    return genere != '0';
}

File: CodiceFiscaleJava/codicefiscalePkg/Sesso.java
package codicefiscalePkg;


/**
 * La classe Sesso rappresenta il genere
 * di una Persona. Contiene funzioni utili per il controllo e
 * il calcolo del codice fiscale e del suo inverso.
 */
public class Sesso extends DatiPersona {
	private char genere;

	public Sesso() { this('0'); }

	public Sesso(char s) { genere = s; }
	
	
	//costruttore di copia
	public Sesso(final Sesso x) {
		genere = x.genere;
	}

	
	/** @return carattere che rappresenta il genere */
	public char estraiGenere() {
		return genere;
	}
		
	
	/**
	 * @param cod contiene sottostringa del codice fiscale che identifica giorno di nascita e sesso.
	 * Se valore compreso tra 0 e 31 allora genere maschile,
	 * altrimenti genere femminile.
	 */
	public void creaDaCodice (final String cod) throws CFexception {
	    //validita' data (e quindi sesso) controllata nei precedenti controlli su Data
	
	    //estrazione codice giorno da codice fiscale (sulla base del quale e' calcolato il sesso)
		int n = Integer.parseInt(cod);
	    if(n>0 && n<=31) genere = 'M';
	        else genere = 'F'; //da metodo Nascita.creaDaCodice, l'altro intervallo e' necessariamente n>40&&n<=71
	}


	/**
	 * @return "40" se genere femminile, "0" se genere maschile.
	 * Rappresenta il valore da sommare a giorno di nascita
	 * nel codice fiscale.
	 */
	public String generaCodice() {
	    if(genere == 'f' || genere == 'F') return "40";
	    else return "0";
	}

	
	/** @throws CFexception se genere diverso da m, oppure M, oppure f, oppure F. */
	public void controllo () throws CFexception {
	    if(genere!='m' && genere!='f' && genere!='M' && genere!='F') throw new CFexception("Sesso non corretto.");
	}
	
	
	/**
	 * overrides java.lang.Object.toString
	 * @return stringa descrittiva del campo dati di Sesso
	 */
	public String toString() {
	    if(genere == 'M' || genere == 'm') return "MASCHIO";
	    else return "FEMMMINA";
	}
}